src/clone.c

```diff
@@ -53,6 +56,76 @@
 # define ARG_CTID	4
 #endif
 
+static bool show_namespace = false;
+extern void namespace_auxstr_init(void)
+{
+	show_namespace = true;
+}
+
+/* derrived from print_dirfd */
+static const char *
+read_namespace_id(int pid, const char *const ns_type)
+{
+	static const char ns_path[] = "/proc/%d/ns/%s";
+	/* "cgroup" is the longest name in the dentries in /proc/$pid/ns. */
+	char linkpath[sizeof(ns_path) + sizeof(int) * 3 + sizeof("cgroup")];
+	xsprintf(linkpath, ns_path, pid, ns_type);
+
+	static char buf[PATH_MAX + 1];
+	ssize_t n = readlink(linkpath, buf, sizeof(buf));
+	if ((size_t) n >= sizeof(buf))
+		return NULL;
+	buf[n] = '\0';
+	return buf;
+}
+
+static const char *
+get_namespace_auxstr(int pid, uint64_t flags)
+{
+	static char str[sizeof("cgroup:[4026531835], "
+			          "ipc:[4026531839], "
+			          "mnt:[4026531841], "
+			          "net:[4026531840], "
+			          "pid:[4026531836], "
+			         "time:[4026531834], "
+			         "user:[4026531837], "
+			          "uts:[4026531838]" )
+			/* for the last null char */
+			+ 1];
+	char *p = str;
+
+#define APPEND(NS_TYPE) do {				\
+		const char *ns_name = read_namespace_id(pid, NS_TYPE); \
+		if (ns_name == NULL)				       \
+			break;				\
+		if (str != p)				\
+			p = xappendstr(str, p, ", ");	\
+		p = xappendstr(str, p, "%s", ns_name);	\
+	} while (0)
+
+	*p = '\0';
+	if (flags & CLONE_NEWCGROUP)
+		APPEND("cgroup");
+	if (flags & CLONE_NEWIPC)
+		APPEND("ipc");
+	if (flags & CLONE_NEWNS)
+		APPEND("mnt");
+	if (flags & CLONE_NEWNET)
+		APPEND("net");
+	if (flags & CLONE_NEWPID)
+		APPEND("pid");
+	if (flags & CLONE_NEWTIME)
+		APPEND("time");
+	if (flags & CLONE_NEWUTS)
+		APPEND("uts");
+	if (flags & CLONE_NEWUSER)
+		APPEND("user");
+
```

Comment:
Wouldn't it be better to iterate over an array of pairs `{ flag, string }`? e.g.
```
struct {
	uint64_t flags;
	const char *str;
} ns[] = {
	{ CLONE_NEWCGROUP | CLONE_INTO_CGROUP, "cgroup" },
	{ CLONE_NEWIPC, "ipc" },
	{ CLONE_NEWNS, "mnt" },
	{ CLONE_NEWNET, "net" },
	{ CLONE_NEWPID, "pid" },
	{ CLONE_NEWTIME, "time" },
	{ CLONE_NEWUTS, "uts" },
	{ CLONE_NEWUSER, "user" },
};

*p = '\0';
for (size_t i = 0; i < ARRAY_SIZE(ns); ++i) {
	if (flags & ns[i].flags) {
		const char *ns_name = read_namespace_id(pid, ns[i].str);
		if (ns_name) {
			if (str != p)
				p = xappendstr(str, p, ", ");
			p = xappendstr(str, p, "%s", ns_name);
		}
	}
}
```

Reply:
(Locally fixed.)